commit b773c5bd9c77b9c6a3598cc43201acc062986d38
Author: Olivier Fourdan <ofourdan@redhat.com>
Date:   Tue Mar 6 11:08:40 2012 +0100

    wacom: Update from gnome-settings-daemon

diff --git a/panels/wacom/gsd-wacom-rr-helpers.c b/panels/wacom/gsd-wacom-rr-helpers.c
index facd972..ce2906e 100644
--- a/panels/wacom/gsd-wacom-rr-helpers.c
+++ b/panels/wacom/gsd-wacom-rr-helpers.c
@@ -70,7 +70,7 @@ gsd_rr_check_xrandr (GdkScreen *gdk_screen,
 		int major_version, minor_version;
 
 		XRRQueryVersion (dpy, &major_version, &minor_version);
-		if (major_version < 2 && minor_version < 2) {
+		if (major_version < 1 && minor_version < 2) {
 			g_set_error (error, GNOME_RR_ERROR, GNOME_RR_ERROR_NO_RANDR_EXTENSION,
 				     _("RANDR extension is too old (%i.%i < 1.2)"), major_version, minor_version);
 			return FALSE;
